File: examples/java/gmsub_disp_rr.java
/*
 * Copyright 2007-2016 United States Government as represented by the
 * Administrator of The National Aeronautics and Space Administration.
 * No copyright is claimed in the United States under Title 17, U.S. Code.
 * All Rights Reserved.
 */



/** 
 * @file gmsub_disp.java
 *
 * A simple example Java GMSEC subscriber using Dispatch and request/reply
 *
 */

import gov.nasa.gsfc.gmsec.api.*;
import gov.nasa.gsfc.gmsec.api.util.Log;

import java.util.ArrayList;


class PublishDispatch extends Callback
{
	public void onMessage(Connection conn, Message msg)
	{
		Log.info("[PublishDispatch.onMessage] Received:\n" + msg.toXML());
	}
}


class DispReplyCallback extends ReplyCallback
{
	public void onReply(Connection conn, Message request, Message reply)
	{
		Log.info("[DispReplyCallback.onReply] Received Reply:\n" + reply.toXML());
	}

	public void onEvent(Connection conn, Status status, Connection.ConnectionEvent event)
	{
		if (status.isError())
		{
			Log.warning("[DispReplyCallback.onEvent] Status=" + status.get());
		}

		Log.info("Press the enter key to exit");
	}
};


public class gmsub_disp_rr implements Example
{
	Config     config;
	Connection connection;
	Connection connectionADisp;
	Message    request;
	ArrayList<SubscriptionInfo> info = new ArrayList<SubscriptionInfo>();


	gmsub_disp_rr(Config config) throws ExampleException
	{
		this.config = config;

		Util.initialize(config);

		if (!Util.hasConnectionType(config))
		{
			printUsage();
			throw new ExampleException("Invalid option");
		}
	}


	private void printUsage()
	{
		System.err.println("\nusage: java gmsub_disp_rr connectionType=<middleware> "
		                   + "[ <parameter>=<value> ]\n"
		                   + "\n\tNote that the parameter 'connectiontype' is required. "
		                   + "\n\tThe rest of other parameters are optional.\n"
		                   + "\n\tserver=<server name> "
		                   + "(required if middleware is not bolt/MB locally)"
		                   + "\n\tsubject=<subject name>"
		                   + "\n\tsubject.<N>=<sub-subject name>"
		                   + "\n\t\tNote: N must be 1 or greater"
		                   + "\n\t\tFor example, subject.1=GMSEC.A subject.2=GMSEC.B"
		                   + "\n\tmsg_timeout_ms=<timeout period (milliseconds)>" 
		                   + "\n\tprog_timeout_s=<timeout period (seconds)>"  
		                   + "\n\tloglevel=<log level>"
		                   + "\n\tcfgfile=<config_filepath.xml>"
		                   + "\n\nFor more information, see API User's Guide\n\n");
	}


	public boolean run()
	{
		boolean result = true;

		// determine what subjects to listen to
		java.util.List<String> subjects = new java.util.ArrayList<String>(10);

		String value = Util.get(config, "SUBJECT");
		if (value != null) {
			subjects.add(value);
		}
		else {
			for (int i = 1;; ++i) {
				String tmp = Util.get(config, "SUBJECT." + i);
				if (tmp != null)
					subjects.add(tmp);
				else
					break;
			}
		}

		if (subjects.isEmpty()) {
			String subject = Util.get(config, "SUBJECT", "GMSEC.>");
			subjects.add(subject);
		} else {
			subjects.add("GMSEC.TERMINATE");
		}
		
		int msg_timeout_ms = Util.get(config, "MSG_TIMEOUT_MS", gmsecAPI.WAIT_FOREVER);
		int prog_timeout_s = Util.get(config, "PROG_TIMEOUT_S", gmsecAPI.WAIT_FOREVER);
		
		// Output GMSEC API version
		Log.info(Connection.getAPIVersion());

		try
		{
			// Establish a connection
			connection = Util.openConnection(config);

			// Output middleware version
			Log.info("Middleware version = " + connection.getLibraryVersion());

			// Subscribe
			for (String subject : subjects)
			{
				Log.info("subscribing to '" + subject + "' with PublishDispatch");
				info.add(connection.subscribe(subject, new PublishDispatch()));
			}

			Log.info("Starting Auto-Dispatch");
			connection.startAutoDispatch();

			// Create request message subject
			String subject = Util.get(config, "SUBJECT", "GMSEC.TEST.REQUEST");

			// Create request message
			Message request = new Message(subject, Message.MessageKind.REQUEST);

			// Add fields to the request message
			request.addField("QUESTION", "Does the request/reply functionality still work?");
			request.addField("COMPONENT", "GMSUB-DISP-RR");

			// Display XML representation of the request message
			Log.info("Prepared Request Message:\n" + request.toXML());

			// Issue request
			connection.request(request, msg_timeout_ms, new DispReplyCallback(), 0);

			// Wait
			int deltaTime = 0;
			if (prog_timeout_s == gmsecAPI.WAIT_FOREVER)
			{
				Log.info("Press the enter key to exit");
				try {
					while(System.in.available() == 0)
					{
						Util.millisleep(100);
					}
				}
				catch (Exception e) {
				}
			}
			else
			{
				Util.millisleep(prog_timeout_s * 1000);
				Log.info("Program Timeout Reached");
			}

			Log.info("Stopping Auto-Dispatch");
			connection.stopAutoDispatch();
		}
		catch (GMSEC_Exception e)
		{
			Log.error("GMSEC_Exception: " + e.toString());
			result = false;
		}
		finally
		{
			try
			{
				cleanup();
			}
			catch (GMSEC_Exception e)
			{
				Log.error("GMSEC_Exception: " + e.toString());
				result = false;
			}
		}

		return result;
	}


	public boolean cleanup() throws GMSEC_Exception
	{
		if (connection != null)
		{
			for(int i = info.size()-1; i >= 0; i-- )
			{
				Log.info("Unsubscribing from " + info.get(i).getSubject());
				connection.unsubscribe(info.get(i));
				info.remove(i);
			}
			Util.closeConnection(connection);
		}

		return true;
	}


	public static void main(String args[])
	{
		try
		{
			Config config = new Config(args);

			Util.addToConfigFromFile(config);

			new gmsub_disp_rr(config).run();
		}
		catch (ExampleException e)
		{
			System.err.println(e.toString());
		}
	}
}

File: examples/perl/gmsub_disp_rr.pl
#
# Copyright 2007-2016 United States Government as represented by the
# Administrator of The National Aeronautics and Space Administration.
# No copyright is claimed in the United States under Title 17, U.S. Code.
# All Rights Reserved.
#

# 
# gmsub_disp_rr.pl
# 
# This file contains an example subscriber using the GMSEC API dispatcher.
# The callback is registered with the Connection.  A separate thread, transmits a request
# and waits for a reply.  The Dispatcher is then started, which launches in its own
# thread.
# 


use strict;
use Util;

use libgmsec_perl;

*isa = \&UNIVERSAL::isa;

# PublishCallback Callback Package
{
	package PublishCallback;
	use base 'libgmsec_perl::Callback';

	sub onMessage
	{
		my ($self, $conn, $message) = @_;

		libgmsec_perl::LogInfo("[PublishCallback::onMessage] Received:\n" . $message->toXML());
	}
} # end PublishCallback definition


# RequestCallback Callback Package
{
	package RequestCallback;
	use base 'libgmsec_perl::ReplyCallback';

	sub onReply
	{
		my ($self, $conn, $message, $reply) = @_;

		libgmsec_perl::LogInfo("[ReplyCallback::onMessage] Received Reply:\n" . $reply->toXML());
	}

	sub onEvent
	{
		my ($self, $status, $event) = @_;

		libgmsec_perl::LogInfo("[ReplyCallback::onEvent] Status=\n" . $status->get()
					. " event=" . $event);
	}
} # end RequestCallback definition


{
	my $task = bless({});
	Util::driver($task, @ARGV);
}


sub printUsage
{
	my ($self, $config) = @_;
	my $hasConnectionType = Util::hasConnectionType($config);
	return 0 if ($hasConnectionType);

	print "\nusage: perl gmsub_disp_rr.pl connectiontype=<middleware> [ <parameter>=<value> ]" .

	"\n\n\tNote that the parameter 'connectiontype' is required." .
	"\n\tThe rest of other parameters are optional." .

	"\n\n\tserver=<server name> (required if middleware is not bolt/MB locally)" .
	"\n\tsubject=<subject name>" .
	"\n\tsubject.N=<sub-subject name>" .
	"\n\t\tNote: N must be 1 or greater" .
	"\n\t\tFor example, subject.1=GMSEC.A subject.2=GMSEC.B"  .
	"\n\tcfgfile=<config_filename.xml>" .
	"\n\tmsg_timeout_ms=<message timeout period (milliseconds)>" .
	"\n\tprog_timeout_s=<program timeout period (seconds)>" .
	"\n\nFor more information, see API User's Guide\n";

	return 1;
}


sub run
{
	my ($self, $config) = @_;

	# Determine run-time settings
	my @subjects = Util::determineSubjects($config);
	my $msg_timeout_ms = Util::getInteger($config, 'MSG_TIMEOUT_MS', $libgmsec_perl::GMSEC_WAIT_FOREVER);
	my $prog_timeout_s = Util::getInteger($config, 'PROG_TIMEOUT_S', $libgmsec_perl::GMSEC_WAIT_FOREVER);

	# Output GMSEC API version
	libgmsec_perl::LogInfo(libgmsec_perl::Connection::getAPIVersion());

	my $connection;

	eval
	{
		$config->addValue("isThreaded", "true");

		# Create the Connection
		$connection = libgmsec_perl::Connection::create($config);
		$self->{CONNECTION} = $connection;

		# Connect
		$connection->connect();

		# Output middleware version
		libgmsec_perl::LogInfo("Middleware version = " . $connection->getLibraryVersion());

		# Output example program-specific information
		libgmsec_perl::LogInfo("Using subjects: " . join(", ", @subjects));
		if ($msg_timeout_ms > $libgmsec_perl::GMSEC_WAIT_FOREVER)
		{
			libgmsec_perl::LogInfo("Message receipt timeout set to occur after " . $msg_timeout_ms . " ms");
		}

		# Subscribe using PublishCallback
		my $condition = libgmsec_perl::Condition->new();
		my $publishCallback = PublishCallback->new();
		$publishCallback->DISOWN();

		# Subscribe
		for my $subject (@subjects)
		{
			libgmsec_perl::LogInfo("Subscribing to " . $subject);
			push(@{$self->{SUBINFO}}, $connection->subscribe($subject, $publishCallback));
		}

		# Register a RequestCallback with the Connection to trigger when a Request times out
		my $requestCallback = RequestCallback->new();
		$connection->registerEventCallback($libgmsec_perl::Connection::REQUEST_TIMEOUT_EVENT, $requestCallback);

		if ($connection->startAutoDispatch())
		{
			libgmsec_perl::LogInfo("Starting AutoDispatcher");
		}
		else
		{
			libgmsec_perl::LogError("Failed to start AutoDispatcher; Exiting the program");
			last; #Break out of the eval statement
		}

		my $requestSubject = "GMSEC.TEST.PUBLISH.NOREPLY";
		if (!($subjects[0] eq "GMSEC.>"))
		{
			$requestSubject = $subjects[0];
		}

		libgmsec_perl::LogInfo("Creating a Request message with subject: " . $requestSubject);
		my $request = libgmsec_perl::Message->new($requestSubject, $libgmsec_perl::Message::REQUEST);
		$request->addField("COMPONENT", "gmsub_disp_rr");

		libgmsec_perl::LogInfo("Sending the Request message");
		$connection->request($request, $msg_timeout_ms, $requestCallback);

		if ($prog_timeout_s == $libgmsec_perl::GMSEC_WAIT_FOREVER)
		{
			libgmsec_perl::LogInfo("Press the enter key to exit");
			my $throwaway = <STDIN>;
		}
		else
		{
			Util::millisleep($prog_timeout_s);
			libgmsec_perl::LogInfo("Program timeout reached!");
		}
	};
	if (isa($@, 'libgmsec_perl::Exception'))
	{
		libgmsec_perl::LogError($@->what());
	}
	elsif($@)
	{
		libgmsec_perl::LogError($@);
	}

	$connection->stopAutoDispatch();

	$self->cleanup();
}


sub cleanup
{
	my ($self) = @_;

	my $connection = $self->{CONNECTION};

	if ($connection)
	{
		for my $subInfo (@{$self->{SUBINFO}})
		{
			eval
			{
				libgmsec_perl::LogInfo("Unsubscribing from " . $subInfo->getSubject());
				$connection->unsubscribe($subInfo);
			};
			if (isa($@, 'libgmsec_perl::Exception'))
			{
				libgmsec_perl::LogError($@->what());
			}
			elsif($@)
			{
				libgmsec_perl::LogError($@);
			}
		}

		libgmsec_perl::Connection::destroy($connection);

	}

	libgmsec_perl::Connection::shutdownAllMiddlewares();
}


1;

__END__
